### PR TITLE
[Training/Python] Add option to enable symbolic shape inference

### PR DESCRIPTION
**Pull request by KeDengMS**
and allow save model to directory

**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py [hunk 1]
```diff
@@ -198,23 +202,16 @@ def _preprocess(self, in_mp):
 
         # returns True if no more ready nodes
         def _insert_ready_nodes():
-            ready_nodes = [pn for pn in pending_nodes if all([i in defined for i in pn.input if i])]
+            ready_nodes = [pn for pn in pending_nodes if len(pn.input) == 0 or all([i in defined for i in pn.input if i])]
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py:205]: >len(pn.input) == 0  [](start = 57, length = 19)

nit:

if not pn.input

Comment (KeDengMS) [reply onnxruntime/core/providers/nuphar/scripts/symbolic_shape_infer.py]: Thanks, rewritten this part

---
In reply to: [487200955](https://github.com/microsoft/onnxruntime/pull/5107#discussion_r487200955) [](ancestors = 487200955)

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -681,6 +697,18 @@ def _init_session(self):
             return
 
         self._verify_fully_optimized_model(self.onnx_model_)
+
+        if self.save_model_dir or self.run_symbolic_shape_infer:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:701]: What scenarios are being unlocked by this change on ORTTrainer? Should this be done for both training and evaluation? InferenceSession is handled by https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/session.py#L158

**Review by thiagocrepaldi — COMMENTED** [review]
@spandantiwari FYI

File: orttraining/orttraining/python/ort_trainer.py (hunk 2 above)
Comment (KeDengMS) [reply orttraining/orttraining/python/ort_trainer.py:701]: BART needs this change, so most tensors could be statically planned. Inference does not rely on shape info for memory pattern though. ([https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/framework/session_state.cc#L313-L352](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/framework/session_state.cc#L313-L352))

---
In reply to: [488801960](https://github.com/microsoft/onnxruntime/pull/5107#discussion_r488801960) [](ancestors = 488801960)

**Review by liqunfu — APPROVED** [review]
Eventually we want merge shape inference function into ort c++ side.

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
Looks good to me, great to have this as an option!

**Pull request merged**